pkg/controller/node/controller_utils.go

```diff
@@ -284,7 +323,7 @@ func recordNodeStatusChange(recorder record.EventRecorder, node *api.Node, new_s
 // terminatePods will ensure all pods on the given node that are in terminating state are eventually
 // cleaned up. Returns true if the node has no pods in terminating state, a duration that indicates how
 // long before we should check again (the next deadline for a pod to complete), or an error.
-func terminatePods(kubeClient clientset.Interface, recorder record.EventRecorder, nodeName string, since time.Time, maxGracePeriod time.Duration) (bool, time.Duration, error) {
+func terminatePods(kubeClient clientset.Interface, recorder record.EventRecorder, nodeName string, nodeStatus nodeStatusData, since time.Time, maxGracePeriod time.Duration) (bool, time.Duration, error) {
```

Comment:
There is the edge case where the node is deleted and the controller is restarted.  Although if the pod garbage collector is maintaining that invariant we should be ok, but if we're still observing that in this controller then we have to follow the same rules as graceful deletion:

1. The node controller maintains its own clock, and makes decisions based on that clock, not on external timestamps
2. We *must* wait at least forgiveness period on that clock before triggering deletion, just like graceful deletion and forcible termination (the node and namespace controllers will always wait at least T seconds before force deleting a pod, preserving the guarantee that the pod won't be terminated in less than T seconds).